sdk/python/feast/infra/online_stores/astra.py

```diff
@@ -0,0 +1,297 @@
+        for entity_key, values, timestamp, created_ts in data:
+            entity_key_bin = serialize_entity_key(entity_key)
+            timestamp = _to_naive_utc(timestamp)
+            if created_ts is not None:
+                created_ts = _to_naive_utc(created_ts)
+            for feature_name, val in values.items():
+                table_name = _table_id(project, table)
+                keyspace = config.online_store.keyspace
+
+                # Now create Insert command
+                insert_cql = _create_cql_insert_record(keyspace,
+                                                       table_name,
+                                                       column_names=["entity_key",
+                                                                     "feature_name",
+                                                                     "value",
+                                                                     "event_ts",
+                                                                     "created_ts"])
+                values = [entity_key_bin,
+                          feature_name,
+                          val.SerializeToString(),
+                          timestamp,
+                          created_ts]
+                self._session = self._initialize_astra_session(config.online_store)
+                prepared = self._session.prepare(insert_cql)
+                self._session.execute(
+                    prepared.bind(values))
+
+    def online_read(
+            self,
+            config: RepoConfig,
+            table: Union[FeatureTable, FeatureView],
+            entity_keys: List[EntityKeyProto],
+            requested_features: Optional[List[str]] = None,
+    ) -> List[Tuple[Optional[datetime], Optional[Dict[str, ValueProto]]]]:
+
+        result: List[Tuple[Optional[datetime], Optional[Dict[str, ValueProto]]]] = []
+        project = config.project
+        table_name = _table_id(project, table)
+        keyspace = config.online_store.keyspace
+
+        all_rows = []
+        for entity_key in entity_keys:
+            entity_key_bin = serialize_entity_key(entity_key)
+            cql_query = _create_select_cql(keyspace, table_name,
+                                           column_to_select=["feature_name",
+                                                             "value",
+                                                             "event_ts"],
+                                           conditions_eq={"entity_key": entity_key_bin})
+            self._session = self._initialize_astra_session(config.online_store)
+            prepared_statement = self._session.prepare(cql_query)
+
+            all_rows += self._session.execute(
+                prepared_statement.bind([entity_key_bin])
+                                 ).all()
+
+        # Now find the result
+
+        for row in all_rows:
+            res = {}
+            feature_name = row.feature_name
+            value = row.value

```

Comment:
what happens if the cql fails for an individual row? e.g. the data isn't found.

Reply:
Actually, .all function will return empty list if there will be no records for that specific query. So in other cases like connection issue then there will be exception like other codes, I go through all the online stores how they are handling and I sync with them